Optimzing merge all

guydou opened the pull request:
using the following changes I was able to reduce the time it takes to `merge_all` of 4K rasters to 1.5 sec after it was 20 sec

I think I can achieve more

guydou committed:
footprint is calcualted out of shape, affine and resoultion, if all the
3 are identical there is no need to check footprint

footprint calculation takes a long while

guydou committed:
limit_to_bands also copies the GeoRaster which takes a while

I interduce subimage, that returns the image wihout the need to copy
anything

guydou committed:
one thing that take a long while is copying the data each of the
iteration

but in the end we copy the new image during the iteration so we don't
need to copy again everything

so I suggest to start merge_all with a fresh copy of empty and to apply
to it the changes, it should have already the right affine, resolution
crs

guydou committed:
allow footprint caching so we will not calculate over and over again

astrojuanlu commented on telluric/georaster.py:
```diff
@@ -1076,9 +1089,17 @@ def bounds(self):
+            coords.append([shape.x, shape.y])
+
+        shp = Polygon(coords)
+        self._footprint = GeoVector(shp, self.crs)
```
What about using `GeoVector.from_bounds`?

guydou replied:
I just copied that was there in the first place, the only thing I added is just caching the product of the calculation 

so the question is to you, I guess

astrojuanlu replied:
Fair enough, could you please add a `# TODO` here so I don't forget? I'll do it after we merge this pull request.

arielze replied:
it's about time :)

astrojuanlu commented on telluric/georaster.py:
```diff
@@ -317,8 +326,7 @@ class GeoRaster2(WindowMethodsMixin):
     """
-
-    def __init__(self, image=None, affine=None, crs=None, filename=None, band_names=None, nodata=0, shape=None):
+    def __init__(self, image=None, affine=None, crs=None, filename=None, band_names=None, nodata=0, shape=None, footprint=None):
```
Are we going to protect ourselves against inconsistent input? I imagine `affine`, `crs`, `shape` and `footprint` can overdetermine the image.

astrojuanlu commented on telluric/georaster.py:
```diff
@@ -103,16 +103,16 @@ def merge_all(rasters, roi, dest_resolution=None, merge_strategy=MergeStrategy.U
         roi, resolution=dest_resolution, band_names=rasters[0].band_names,
         dtype=rasters[0].dtype)
 
+    # empty = rasters[0]
```
Leftover?

astrojuanlu reviewed (changes requested):
Left some comments, and also there are some PEP8 errors https://travis-ci.org/satellogic/telluric/jobs/383197579#L755

guydou committed:
deviding the GeoRaster2 to GeoRaster2 and Raster

Raster will have the methematical datamemebers such as image and
band_names

While GeoRaster2 has the Geo references

I am using this object in the _merge function to remove all the copies

guydou committed:
style changes

guydou committed:
code style

guydou committed:
fixing mypy

guydou committed:
coding style

astrojuanlu commented on \:
```diff
@@ -0,0 +1,1561 @@
+import json
```
Extra file?

guydou replied:
I don't understand this comment

guydou replied:
I don't get this comment

astrojuanlu commented on telluric/georaster.py:
```diff
@@ -303,39 +302,20 @@ class GeoRaster2NotImplementedError(GeoRaster2Error, NotImplementedError):
 
 
-class GeoRaster2(WindowMethodsMixin):
+class Raster():
```
What about making `_Raster` private?

astrojuanlu reviewed (changes requested):
Overall this looks good to me, just left some minor comments.

arielze commented on telluric/georaster.py:
```diff
@@ -395,6 +375,60 @@ def _set_bandnames(self, band_names=None):
 
+    def subimage(self, bands):
+        if isinstance(bands, str):
+            bands = [bands]
```
I think `bands = bands.split(',')` is better

arielze commented on telluric/georaster.py:
```diff
@@ -395,6 +375,60 @@ def _set_bandnames(self, band_names=None):
+            bands = [bands]
+
+        missing_bands = set(bands) - set(self.band_names)
+        if missing_bands:
```
so you raise if not all bands exists, you could return only what exists and empty arrays for the non existing bands, or just what exists.

guydou replied:
That was the behavior in `limit_to_bands` I just copied it from the original method,  I didn't want to change its behavior.

I think this disscussion should be made on different thread

arielze commented on telluric/georaster.py:
```diff
@@ -395,6 +375,60 @@ def _set_bandnames(self, band_names=None):
+
+        bands_indices = [self.band_names.index(band) for band in bands]
+        subimage = self.image[bands_indices, :, :]
+        return subimage
```
I understand why you didn't return `subimage.copy()` but this could be dangerous.
1. the values of the  return subimage can be changed from the source raster.
2. the source raster image can mistakenly be change by the receiver.
3. if the reference to the receiver can prevent GC of the source raster == memory issues

guydou replied:
we have the same problem with `raster.image` why there it isn't a problem and here it is, the idea behind this method is to do the same as `.image` but slice by bands 

I don' want to copy, since that is the whole point of optimzing, and in the context of what I am doing it okay since there is no direct equality

arielze commented on telluric/georaster.py:
```diff
@@ -1018,7 +1052,7 @@ def _repr_png_(self):
         return self.to_png(transparent=True, thumbnail_size=512, resampling=Resampling.nearest, stretch=True)
 
-    def limit_to_bands(self, bands):
+    def subimage(self, bands):
```
why did you had to override this method? isn't it the same as for `Raster`

guydou replied:
leftovers

arielze commented on telluric/georaster.py:
```diff
@@ -103,16 +103,16 @@ def merge_all(rasters, roi, dest_resolution=None, merge_strategy=MergeStrategy.U
-                   merge_strategy=merge_strategy)  # type: Callable[[GeoRaster2, GeoRaster2], GeoRaster2]
-    return reduce(func, rasters, empty)
+    projected_rasters = [_prepare_other_raster(empty, raster) for raster in rasters]
+    projected_rasters = [raster for raster in projected_rasters if raster is not None]
```
I think you can merge this with the previous loop

guydou replied:
to do that I will need to run `_prepare_other_raster` twice, it is the reason for existence of NONE,  the second loop doesn't cost anything, and it is more clear like that

arielze commented on telluric/georaster.py:
```diff
@@ -103,16 +103,16 @@ def merge_all(rasters, roi, dest_resolution=None, merge_strategy=MergeStrategy.U
-    func = partial(merge_to_first,
-                   merge_strategy=merge_strategy)  # type: Callable[[GeoRaster2, GeoRaster2], GeoRaster2]
-    return reduce(func, rasters, empty)
+    projected_rasters = [_prepare_other_raster(empty, raster) for raster in rasters]
```
Nice!
it will be interesting to implement this when we would like to do it lazy

arielze commented on telluric/georaster.py:
```diff
@@ -206,27 +207,24 @@ def _merge(one, other, merge_strategy=MergeStrategy.UNION):
         new_image = np.ma.MaskedArray(
             np.concatenate(all_data),
-            mask=new_mask[None].repeat(len(new_bands), axis=0) == 1
+            mask=[new_mask]*len(new_bands)
```
are you sure this is what you where looking for, this will keep a reference to `new_mask` the `len(new_bands)` times.
`mask[0] is mask[1]` will return `True`
`new_mask is mask[n]` will be `True` for any n

guydou replied:
can you think of a better way to do that, the original way costs 3 times more per raster

arielze commented on telluric/georaster.py:
```diff
@@ -125,19 +125,20 @@ def _prepare_other_raster(one, other):
 
 
-def _merge(one, other, merge_strategy=MergeStrategy.UNION):
+def _merge(one, other, merge_strategy=MergeStrategy.UNION, common_bands=None):
```
can you explain why you had to add the band_names argument ?

guydou replied:
I am using it in the line 194, 
Before the change, we did it twice, in line 194 and right afterward because of the recursion also [here](https://github.com/satellogic/telluric/pull/41/files#diff-6dcac23673708582303bd00fc00849e4L140).

Maybe I should change the name to requested_bands

arielze commented on telluric/georaster.py:
```diff
@@ -206,27 +207,24 @@ def _merge(one, other, merge_strategy=MergeStrategy.UNION):
-
-        return res_common.copy_with(image=new_image, band_names=new_bands)
-
+        return Raster(image=new_image, band_names=new_bands)
```
I think here you should do `return Raster(image=deepcopy(new_image), band_names=deepcopy(new_bands))` o breake reference to `one` and `other` objects

guydou replied:
why? they are created in the lines before

arielze reviewed (commented):
did you check that `merge_all` tests cover situations where merged rasters are of different shapes and are partially and not intersecting with the roi?

guydou committed:
removing a method we didn't have to override

guydou committed:
better code style

guydou committed:
refactor the Raster class to be private

guydou committed:
fixing mypy

guydou committed:
fixing the test by doing a split only for strings

guydou committed:
fixing style

guydou committed:
Merge branch 'master' into optimzing_merge_all

arielze reviewed (approved):
good job @guydou !

guydou committed:
adding a warnning comment about using refereces

guydou committed:
requested_bands instead of common_bands

guydou committed:
Merge branch 'optimzing_merge_all' of https://github.com/satellogic/telluric into optimzing_merge_all

guydou committed:
delete unnecessary file

guydou committed:
reverting a mistake

guydou committed:
complete common_band change

guydou committed:
changin references to rasterio documentation

astrojuanlu reviewed (approved):
Well done @guydou!
